src/main/java/com/onelogin/saml/Utils.java

```diff
@@ -146,52 +163,207 @@ public static Document loadXML(String xml) throws Exception {
-	 		for (int i=0; i<nodeList.getLength() ; i++) {
-	 			Element elem = (Element) nodeList.item(i);
-	 			Attr attr = (Attr) elem.getAttributes().getNamedItem("ID");
-	 			elem.setIdAttributeNode(attr, true);
-	 		}
+			XPath xpath = XPathFactory.newInstance().newXPath();
+			XPathExpression expr = xpath.compile("//*[@ID]");
+			NodeList nodeList = (NodeList) expr.evaluate(doc, XPathConstants.NODESET);
+			for (int i=0; i<nodeList.getLength() ; i++) {
+				Element elem = (Element) nodeList.item(i);
+				Attr attr = (Attr) elem.getAttributes().getNamedItem("ID");
+				elem.setIdAttributeNode(attr, true);
+			}
 			return doc;
 		} catch (Exception e) {
-			e.printStackTrace();
+			log.error("Error executing loadXML: " + e.getMessage(), e);
+		}
+		return null;
+	}
+
+	/**
+	 * This function attempts to validate an XML against the specified schema.
+	 * 
+	 * @param xml The XML document which should be validated
+	 * @param schema The schema filename which should be used
+	 * @throws Exception
+	 */
+	public static Document validateXML(Document xml, String schemaName)
+			throws Exception {
+		return validateXML(getStringFromDocument(xml), schemaName);
+	}
+
+	/**
+	 * This function attempts to validate an XML string against the specified
+	 * schema. It will parse the string into a DOM document and validate this
+	 * document against the schema.
+	 * 
+	 * @param xml The XML document which should be validated
+	 * @param schema The schema filename which should be used
+	 * @throws Exception
+	 */
+	public static Document validateXML(String xmlString, String schemaName, Boolean...debugMode)
+			throws Exception {
+
+		try {
+			String schemaFullPath = "schemas" + File.separatorChar + schemaName;
+			log.debug("schemaFullPath: " +schemaFullPath);
+			ClassLoader classLoader = Utils.class.getClassLoader();
+			URL schemaFile = classLoader.getResource(schemaFullPath);
+			SchemaFactory schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+			Schema schema = schemaFactory.newSchema(schemaFile);
+			Validator validator = schema.newValidator();
+			XMLErrorHandler errorHandler = new XMLErrorHandler();
+			validator.setErrorHandler(errorHandler);
+			validator.validate(new StreamSource(new StringReader(xmlString)));
+
+			if (errorHandler.getErrorXML().size() > 0) {
+				throw new Error("Invalid XML. See the log");
+			}
+		} catch (Error e) {
+			throw e;
+		} catch (Exception ex) {
+			log.error("Error executing validateXML: " + ex.getMessage(), ex);
+			throw ex;
+		}
+		return convertStringToDocument(xmlString);
+	}
+
+	/**
+	 * Validate signature (Message or Assertion).
+	 * 
+	 * @param xml The element we should validate
+	 * @param cert The pubic cert
+	 * @param fingerprint The fingerprint of the public cert
+	 * @return True if the sign is valid, false otherwise.
+	 */
+	public static boolean validateSign(Node signatureElement, Certificate cert, String ...fingerprint) 
```

Comment:
mean that are optional like in php fingerprint = null, for now we are not supporting fingerprint but the name method is ready to support it when we have a chance.